dali/python/backend_impl.cc

```diff
@@ -1248,34 +1248,56 @@ PYBIND11_MODULE(backend_impl, m) {
     .def("RunCPU", &Pipeline::RunCPU, py::call_guard<py::gil_scoped_release>())
     .def("RunGPU", &Pipeline::RunGPU)
     .def("Outputs",
-        [](Pipeline *p) {
+        [](py::object &p_o) {
+          Pipeline *p = py::cast<Pipeline *>(p_o);
           DeviceWorkspace ws;
           p->Outputs(&ws);
 
-          py::list list;
+          py::tuple tup(ws.NumOutput());
           for (int i = 0; i < ws.NumOutput(); ++i) {
             if (ws.OutputIsType<CPUBackend>(i)) {
-              list.append(&ws.Output<CPUBackend>(i));
+              tup[i] = &ws.Output<CPUBackend>(i);
+              // it is not possible to obtain weekref to the tuple and bound it to the pipeline
+              // object to keep the pipeline object alive as long as a returned tuple is valid,
+              // so instead keep pipeline alive as long as any element of the returned
+              // tuple is alive
+              py::detail::keep_alive_impl(tup[i], p_o);
             } else {
-              list.append(&ws.Output<GPUBackend>(i));
+              tup[i] = &ws.Output<GPUBackend>(i);
+              // it is not possible to obtain weekref to the tuple and bound it to the pipeline
+              // object to keep the pipeline object alive as long as a returned tuple is valid,
+              // so instead keep pipeline alive as long as any element of the returned
+              // tuple is alive
+              py::detail::keep_alive_impl(tup[i], p_o);
             }
           }
-          return py::cast<py::tuple>(list);
+          return tup;
         }, py::return_value_policy::take_ownership)
     .def("ShareOutputs",
-        [](Pipeline *p) {
+        [](py::object &p_o) {
+          Pipeline *p = py::cast<Pipeline *>(p_o);
           DeviceWorkspace ws;
           p->ShareOutputs(&ws);
 
-          py::list list;
+          py::tuple tup(ws.NumOutput());
           for (int i = 0; i < ws.NumOutput(); ++i) {
             if (ws.OutputIsType<CPUBackend>(i)) {
-              list.append(&ws.Output<CPUBackend>(i));
+              tup[i] = &ws.Output<CPUBackend>(i);
+              // it is not possible to obtain weekref to the tuple and bound it to the pipeline
+              // object to keep the pipeline object alive as long as a returned tuple is valid,
+              // so instead keep pipeline alive as long as any element of the returned
+              // tuple is alive
+              py::detail::keep_alive_impl(tup[i], p_o);
             } else {
-              list.append(&ws.Output<GPUBackend>(i));
+              tup[i] = &ws.Output<GPUBackend>(i);
```

Comment:
you can move this outside of the if and have it only once

Reply:
Done